src/vmm/src/guest_config/x86_64/mod.rs

```diff
@@ -43,101 +42,38 @@ pub struct CpuConfiguration {
-                    StaticCpuTemplate::T2CL => static_cpu_templates::t2cl::t2cl(),
-                    StaticCpuTemplate::T2A => static_cpu_templates::t2a::t2a(),
-                    StaticCpuTemplate::None => unreachable!("None state is invalid"),
-                };
-
-                // Include leaves from host that are not present in CPUID template.
-                config.cpuid = template_cpuid
-                    .include_leaves_from(config.cpuid)
-                    .map_err(Error::CpuidJoin)?;
-
-                // TODO: Some MSRs depend on values of other MSRs. This dependency will need to
-                // be implemented. For now we define known dependencies statically in the CPU
-                // templates.
-
-                // Depending on which CPU template the user selected, we may need to initialize
-                // additional MSRs for boot to correctly enable some CPU features. As stated in
-                // the previous comment, we get from the template a static list of MSRs we need
-                // to save at snapshot as well.
-                // C3, T2 and T2A currently don't have extra MSRs to save/set.
-                match template {
-                    StaticCpuTemplate::T2S => {
-                        config.supported_msrs.extend(msr_entries_to_save());
-                        static_cpu_templates::t2s::update_t2s_msr_entries(
-                            &mut config.msr_boot_entries,
-                        );
-                    }
-                    StaticCpuTemplate::T2CL => {
-                        config.supported_msrs.extend(msr_entries_to_save());
-                        static_cpu_templates::t2cl::update_t2cl_msr_entries(
-                            &mut config.msr_boot_entries,
-                        );
-                    }
-                    _ => (),
-                }
-
-                Ok(config)
-            }
-        }
+        Ok(Self { cpuid, msrs })
     }
 
     /// Modifies provided config with changes from template
     pub fn apply_template(self, template: &CustomCpuTemplate) -> Result<Self, Error> {
         let Self {
             mut cpuid,
             mut msrs,
-            supported_msrs,
-            msr_boot_entries,
         } = self;
 
-        let guest_cpuid = match &mut cpuid {
-            Cpuid::Intel(cpuid) => &mut cpuid.0,
-            Cpuid::Amd(cpuid) => &mut cpuid.0,
-        };
+        // Check whether the vendor specified in CPU template matches the actual vendor.
+        let guest_cpuid = match (&mut cpuid, &template.vendor) {
+            (Cpuid::Intel(_), Some(CpuVendor::Amd)) => {
+                Err(Error::CpuVendorMismatched(CpuVendor::Intel, CpuVendor::Amd))
+            }
+            (Cpuid::Amd(_), Some(CpuVendor::Intel)) => {
+                Err(Error::CpuVendorMismatched(CpuVendor::Amd, CpuVendor::Intel))
+            }
+            (Cpuid::Intel(cpuid), _) => Ok(&mut cpuid.0),
```

Comment:
If we genericised our Cpuid type to not be aware of Intel or AMD, would we be able to drop match statements like this, and drop the vendor field from the template?

Reply:
We need to check whether the CPU vendor that the template is intended to be applied on matches the actual CPU vendor, because T2A template should only be applied on AMD and all the other templates should be applied on Intel. If you have other ideas to do this check, it's worth discussing!

Reply:
Hmmm, if validation of static templates is the only use-case here, we should probably consider looking up vendor and validating against the host's CPUID in some way. It shouldn't be necessary for a user to specify the field and value every time, especially which adds a field to aarch64 which does not require it.

Reply:
Fair enough. We already have `src/vmm/src/arch/x86_64/cpu_model.rs`, so we might be able to remove `IntelCpuid` and `AmdCpuid`. Is it okay for you to do this in a follow-up commit along with this change? https://github.com/firecracker-microvm/firecracker/pull/3610#discussion_r1168661090
I guess that this task could become one PR.

Reply:
removed `vendor` field from `CustomCpuTemplate`, instead put vendor check into `get_cpu_template()`.